File: ios/Runner/Cast/CastBridge.swift
import Foundation
import GoogleCast
import os.log

@MainActor
final class CastBridge: NSObject, CastHostApi {

    private let flutterApi: CastFlutterApi
    private let sessionManager: CastSessionManager
    private var currentMedia: MediaInfo?

    init(flutterApi: CastFlutterApi, castContext: GCKCastContext) {
        self.flutterApi = flutterApi
        self.sessionManager = CastSessionManager()
        super.init()

        let googleCastProvider = GoogleCastProvider(castContext: castContext)
        let airPlayProvider = AirPlayProvider()
        sessionManager.registerProvider(googleCastProvider)
        sessionManager.registerProvider(airPlayProvider)

        sessionManager.setStateObserver { [weak self] state in
            self?.handleStateChange(state)
        }
        sessionManager.setDevicesObserver { [weak self] devices in
            self?.flutterApi.onDevicesChanged(devices: devices) { _ in }
        }

        log("Initialized")
    }

    // MARK: - CastHostApi

    func startDiscovery() throws {
        sessionManager.startDiscovery()
    }

    func stopDiscovery() throws {
        sessionManager.stopDiscovery()
    }

    func getDiscoveredDevices() throws -> [CastDevice] {
        sessionManager.getDiscoveredDevices()
    }

    func connect(deviceId: String) throws {
        sessionManager.connect(deviceId: deviceId)
    }

    func disconnect() throws {
        sessionManager.disconnect()
    }

    func showAirPlayPicker() throws {
        sessionManager.showAirPlayPicker()
    }

    func loadMedia(mediaInfo: MediaInfo, autoplay: Bool, positionMs: Int64) throws {
        currentMedia = mediaInfo
        sessionManager.loadMedia(mediaInfo, autoplay: autoplay, positionMs: positionMs)
    }

    func play() throws {
        sessionManager.play()
    }

    func pause() throws {
        sessionManager.pause()
    }

    func seek(positionMs: Int64) throws {
        sessionManager.seek(positionMs: positionMs)
    }

    func stop() throws {
        sessionManager.stop()
        currentMedia = nil
    }

    func setVolume(volume: Double) throws {
        sessionManager.setVolume(volume)
    }

    func setMuted(muted: Bool) throws {
        sessionManager.setMuted(muted)
    }

    nonisolated func dispose() {
        Task { @MainActor in
            log("Disposing...")
            sessionManager.dispose()
        }
    }

    // MARK: - Private

    private func handleStateChange(_ state: SessionSnapshot) {
        let castSessionState = CastSessionState(
            connectionState: state.connectionState,
            playbackState: state.playbackState,
            connectedDevice: state.connectedDevice,
            currentMedia: currentMedia,
            positionMs: state.positionMs,
            durationMs: state.durationMs,
            errorMessage: state.errorMessage
        )
        flutterApi.onStateChanged(state: castSessionState) { _ in }
    }

    private nonisolated func log(_ message: String) {
        os_log("[CastBridge] %{public}@", log: .default, type: .debug, message)
    }
}

File: android/app/src/main/kotlin/dev/rohanjsh/pitcher/cast/CastBridge.kt
package dev.rohanjsh.pitcher.cast

import dev.rohanjsh.pitcher.cast.session.CastSessionManager
import dev.rohanjsh.pitcher.cast.session.SessionSnapshot

class CastBridge(
    private val sessionManager: CastSessionManager,
    private val flutterApi: CastFlutterApi
) : CastHostApi {

    private var currentMedia: MediaInfo? = null

    init {
        sessionManager.setStateObserver { notifyFlutter(it) }
        sessionManager.setDevicesObserver { flutterApi.onDevicesChanged(it) { } }
    }

    override fun startDiscovery() = sessionManager.startDiscovery()
    override fun stopDiscovery() = sessionManager.stopDiscovery()
    override fun getDiscoveredDevices(): List<CastDevice> = sessionManager.getDiscoveredDevices()
    override fun connect(deviceId: String) = sessionManager.connect(deviceId)
    override fun disconnect() = sessionManager.disconnect()
    override fun showAirPlayPicker() { /* No-op on Android */ }

    override fun loadMedia(mediaInfo: MediaInfo, autoplay: Boolean, positionMs: Long) {
        currentMedia = mediaInfo
        sessionManager.loadMedia(mediaInfo, autoplay, positionMs)
    }

    override fun play() = sessionManager.play()
    override fun pause() = sessionManager.pause()
    override fun seek(positionMs: Long) = sessionManager.seek(positionMs)

    override fun stop() {
        sessionManager.stop()
        currentMedia = null
    }

    override fun setVolume(volume: Double) = sessionManager.setVolume(volume)
    override fun setMuted(muted: Boolean) = sessionManager.setMuted(muted)

    fun dispose() = sessionManager.dispose()

    private fun notifyFlutter(state: SessionSnapshot) {
        flutterApi.onStateChanged(
            CastSessionState(
                connectionState = state.connectionState,
                playbackState = state.playbackState,
                connectedDevice = state.connectedDevice,
                currentMedia = currentMedia,
                positionMs = state.positionMs.takeIf { it > 0 },
                durationMs = state.durationMs.takeIf { it > 0 },
                errorMessage = state.errorMessage
            )
        ) { }
    }
}
